js/ui/animation.js

```diff
@@ -99,17 +116,31 @@ var Animation = new Lang.Class({
         this.stop();
     },
 
-    setFrameTimeouts: function(timeoutList) {
-        if (timeoutList.length == 0)
-            throw new Error('Cannot set an empty list as the frame-timeouts for animation!');
+    _parseFrame: function(frameInfo) {
+        if (typeof frameInfo === 'string') {
+            let [frame, delay] = frameInfo.split(' ');
+            frame = parseInt(frame);
 
+            if (delay.indexOf('-') > -1)
+                delay = parseInt(delay);
+
+            return [frame, delay];
+        }
+        return [frameInfo, this._speed];
+    },
```

Comment:
I understand we are passing the metadata format that's already being used in the Clubhouse, however I think this should be parsed by some helper function outside of the `Animation` and set directly it (otherwise we should just pass it the whole json and parse it here too). Also, the more we keep things independent of this arbitrary format, the higher the chances of eventually getting the `Animation` changes upstream.

Reply:
OK! I can do that. So should I extend Animation in the clubhouse? I didn't do it because your previous change to add the different offsets were applied to this class directly, but yeah sounds like this is too much changes for the Animation class.

Reply:
No. I mean, you keep the logic you have here, but **IMO** the `Animation` class shouldn't know about our arbitrary JSON format. It's one thing to have a `setFrameInfo` (similar to what I did with the timeouts), it's another to have it parse the JSON. So I'd parse the JSON in the Clubhouse component (Shell or Clubhouse if we need it only in one place), and pass it to the Animation already with:
`[[frame, delay], [frame, delay], ...]` or whatever is better.
The idea is that the Animation class doesn't know about the string format, just that we need to display the frames at a certain delay.